include/fiction/algorithms/simulation/sidb/ground_state_space.hpp

```diff
@@ -0,0 +1,802 @@
+            clustering.insert(child_clustering.cbegin(), child_clustering.cend());
+        }
+
+        return clustering;
+    }
+
+    template <bound_direction bound>
+    static constexpr inline potential_projection get_proj_bound(const sidb_cluster_ptr& c,
+                                                                const uint64_t          sidb_ix) noexcept
+    {
+        return c->pot_projs.at(sidb_ix).get<bound>();
+    }
+
+    template <bound_direction bound>
+    static constexpr inline double get_next_proj_pot_bound(const sidb_cluster_ptr& c, const uint64_t sidb_ix) noexcept
+    {
+        return c->pot_projs.at(sidb_ix).get_next<bound>().v;
+    }
+
+    template <bound_direction bound>
+    static constexpr inline potential_projection get_proj_state_bound(const sidb_cluster_projector_state& pst,
+                                                                      const uint64_t sidb_ix) noexcept
+    {
+        return pst.cluster->pot_projs.at(sidb_ix).get_pot_proj_for_m_conf<bound>(pst.multiset_conf);
+    }
+
+    static constexpr inline void add_pot_proj(const sidb_cluster_ptr& c, const uint64_t sidb_ix,
+                                              const potential_projection& pp) noexcept
+    {
+        c->pot_projs[sidb_ix].add(pp);
+    }
+
+    static constexpr inline void remove_all_cluster_charge_state_occurrences(const sidb_cluster_projector_state& rm_pst,
+                                                                             const uint64_t sidb_ix) noexcept
+    {
+        rm_pst.cluster->pot_projs[sidb_ix].remove_m_conf(rm_pst.multiset_conf);
+    }
+
+    template <bound_direction bound>
+    static constexpr void update_external_pot_proj_if_bound_removed(const sidb_cluster_projector_state& pst,
+                                                                    const sidb_cluster_receptor_state&  rst) noexcept
+    {
+        const potential_projection& cur_bound = get_proj_bound<bound>(pst.cluster, rst.sidb_ix);
+
+        if (cur_bound.m == pst.multiset_conf)
+        {
+            rst.cluster->update_recv_ext_pot_bound<bound>(
+                rst.sidb_ix, get_next_proj_pot_bound<bound>(pst.cluster, rst.sidb_ix) - cur_bound.v);
+        }
+    }
+
+    constexpr void update_external_potential_projection(const sidb_cluster_projector_state& pst,
+                                                        const sidb_cluster_receptor_state&  rst) const noexcept
+    {
+        update_external_pot_proj_if_bound_removed<bound_direction::LOWER>(pst, rst);
+        update_external_pot_proj_if_bound_removed<bound_direction::UPPER>(pst, rst);
+
+        remove_all_cluster_charge_state_occurrences(pst, rst.sidb_ix);
+    }
+
+    void handle_invalid_state(const sidb_cluster_projector_state& pst) noexcept
+    {
+        // may need to remove the multiset charge configuration from the internal potential projection order too
```

Comment:
is this a TODO?

Reply:
Good catch. I found out that it is not necessary to remove it. Comment can therefore go